From b5b4f4dbb7d585af1e7e733677c7c14434b44f58 Mon Sep 17 00:00:00 2001
From: Jon Hunter <jonathanh@nvidia.com>
Date: Fri, 20 Feb 2026 08:24:06 -0800
Subject: [PATCH 56/59] net: can: mttcan: Fix build for Linux v6.17

In Linux v6.17, the 'const' qualifier for the 'struct cyclecounter'
argument on of the cyclecounter 'read' function pointer was dropped.
Add a test to conftest to detect this and update the Tegra MTT CAN
driver accordingly to fix the build.

Bug 5420210

Upstream-Status: Backport
Change-Id: Idff1429bbf45fe160e16a6dbc51e8cb00909a876
Signed-off-by: Jon Hunter <jonathanh@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-nv-oot/+/3423831
(cherry picked from commit 10569f9bbea4fc8172a94b8d5c5190fc2687b199)
Reviewed-on: https://git-master.nvidia.com/r/c/linux-nv-oot/+/3462468
Reviewed-by: Brad Griffis <bgriffis@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Tested-by: mobile promotions <svcmobile_promotions@nvidia.com>
Reviewed-by: mobile promotions <svcmobile_promotions@nvidia.com>
(cherry picked from commit 76bdc9f5b69e54cd55371ab04f28824f5d028793)
Signed-off-by: Matt Madison <matt@madison.systems>
---
 drivers/net/can/mttcan/hal/m_ttcan.c     |  4 ++++
 drivers/net/can/mttcan/include/m_ttcan.h |  4 ++++
 scripts/conftest/Makefile                |  1 +
 scripts/conftest/conftest.sh             | 18 ++++++++++++++++++
 4 files changed, 27 insertions(+)

diff --git a/drivers/net/can/mttcan/hal/m_ttcan.c b/drivers/net/can/mttcan/hal/m_ttcan.c
index bc2f998f..cbe9e61e 100644
--- a/drivers/net/can/mttcan/hal/m_ttcan.c
+++ b/drivers/net/can/mttcan/hal/m_ttcan.c
@@ -1076,7 +1076,11 @@ int ttcan_controller_init(struct ttcan_controller *ttcan, u32 irq_flag,
 	return 0;
 }
 
+#if defined(NV_CYCLECOUNTER_STRUCT_READ_HAS_CONST_CYCLECOUNTER_ARG)
 u64 ttcan_read_ts_cntr(const struct cyclecounter *ccnt)
+#else
+u64 ttcan_read_ts_cntr(struct cyclecounter *ccnt)
+#endif
 {
 	struct mttcan_priv *priv = container_of(ccnt, struct mttcan_priv, cc);
 
diff --git a/drivers/net/can/mttcan/include/m_ttcan.h b/drivers/net/can/mttcan/include/m_ttcan.h
index 90be8127..a24e9204 100644
--- a/drivers/net/can/mttcan/include/m_ttcan.h
+++ b/drivers/net/can/mttcan/include/m_ttcan.h
@@ -561,5 +561,9 @@ int add_msg_controller_list(struct ttcan_controller *ttcan,
 int add_event_controller_list(struct ttcan_controller *ttcan,
 				struct mttcan_tx_evt_element *txevt,
 				struct list_head *evt_q);
+#if defined(NV_CYCLECOUNTER_STRUCT_READ_HAS_CONST_CYCLECOUNTER_ARG)
 u64 ttcan_read_ts_cntr(const struct cyclecounter *ccnt);
+#else
+u64 ttcan_read_ts_cntr(struct cyclecounter *ccnt);
+#endif
 #endif
diff --git a/scripts/conftest/Makefile b/scripts/conftest/Makefile
index 3131989a..839efaf6 100644
--- a/scripts/conftest/Makefile
+++ b/scripts/conftest/Makefile
@@ -109,6 +109,7 @@ NV_CONFTEST_FUNCTION_COMPILE_TESTS += can_priv_struct_has_struct_data_bittiming_
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += class_attribute_struct_has_const_struct_class_attribute
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += class_create_has_no_owner_arg
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += class_struct_devnode_has_const_dev_arg
+NV_CONFTEST_FUNCTION_COMPILE_TESTS += cyclecounter_struct_read_has_const_cyclecounter_arg
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += define_semaphore_has_number_arg
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += devfreq_dev_profile_has_is_cooling_device
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += devfreq_has_freq_table
diff --git a/scripts/conftest/conftest.sh b/scripts/conftest/conftest.sh
index 9bddb615..c4150f2e 100755
--- a/scripts/conftest/conftest.sh
+++ b/scripts/conftest/conftest.sh
@@ -6883,6 +6883,24 @@ compile_test() {
             compile_check_conftest "$CODE" "NV_CLASS_STRUCT_DEVNODE_HAS_CONST_DEV_ARG" "" "types"
         ;;
 
+        cyclecounter_struct_read_has_const_cyclecounter_arg)
+            #
+            # Determine if the 'cyclecounter' structure 'read' function pointer
+            # has a const 'struct cyclecounter' argument.
+            #
+            # The const qualifier was dropped in Linux v6.17 by commit e78f70bad29c
+            # ("time/timecounter: Fix the lie that struct cyclecounter is const")
+            #
+            CODE="
+            #include <linux/timecounter.h>
+            void conftest(struct cyclecounter *cc) {
+                    u64 (*fn)(const struct cyclecounter *) = cc->read;
+            }"
+
+            compile_check_conftest "$CODE" \
+                    "NV_CYCLECOUNTER_STRUCT_READ_HAS_CONST_CYCLECOUNTER_ARG" "" "types"
+        ;;
+
         define_semaphore_has_number_arg)
             #
             # Determine if the macro DEFINE_SEMAPHORE has a number argument.
-- 
2.43.0

